src/OpenDebugAD7/AD7DebugSession.cs

```diff
@@ -2327,6 +2331,229 @@ protected override void HandleSetBreakpointsRequestAsync(IRequestResponder<SetBr
+            {
+                responder.SetError(new ProtocolException("SetDataBreakpointRequest failed: Missing 'breakpoints'."));
+                return;
+            }
+
+            List<DataBreakpoint> breakpoints = responder.Arguments.Breakpoints;
+            SetDataBreakpointsResponse response = new SetDataBreakpointsResponse();
+
+            Dictionary<string, IDebugPendingBreakpoint2> newBreakpoints = new Dictionary<string, IDebugPendingBreakpoint2>();
+
+            foreach (KeyValuePair<string, IDebugPendingBreakpoint2> b in m_dataBreakpoints)
+            {
+                if (breakpoints.Find((p) => p.DataId == b.Key) != null)
+                {
+                    newBreakpoints[b.Key] = b.Value;    // breakpoint still in new list
+                }
+                else
+                {
+                    b.Value.Delete();   // not in new list so delete it
+                }
+            }
+
+            foreach (DataBreakpoint b in breakpoints)
+            {
+                if (m_dataBreakpoints.ContainsKey(b.DataId))
+                {   // already created
+                    IDebugBreakpointRequest2 breakpointRequest;
+                    if (m_dataBreakpoints[b.DataId].GetBreakpointRequest(out breakpointRequest) == 0 &&
+                                breakpointRequest is AD7BreakPointRequest ad7BPRequest)
+                    {
+                        // Check to see if this breakpoint has a condition that has changed.
+                        if (!StringComparer.Ordinal.Equals(ad7BPRequest.Condition, b.Condition))
+                        {
+                            // Condition has been modified. Delete breakpoint so it will be recreated with the updated condition.
+                            var toRemove = m_dataBreakpoints[b.DataId];
+                            toRemove.Delete();
+                            m_dataBreakpoints.Remove(b.DataId);
+                        }
+                        else
+                        {
+                            if (ad7BPRequest.BindResult != null)
+                            {
+                                response.Breakpoints.Add(ad7BPRequest.BindResult);
+                            }
+                            else
+                            {
+                                response.Breakpoints.Add(new Breakpoint()
+                                {
+                                    Id = (int)ad7BPRequest.Id,
+                                    Verified = true,
+                                    Line = 0
+                                });
+
+                            }
+                            continue;
+                        }
+                    }
+                }
+
+                // Bind the new data bp
+                if (!m_dataBreakpoints.ContainsKey(b.DataId))
+                {
+                    int hr = HRConstants.S_OK;
+
+                    string[] splits = b.DataId.Split(',');
```

Comment:
>string[] splits = b.DataId.Split(','); [](http://example.com/codeflow?start=20&length=38)

You don't want to split it, because the engine provided engine id can contain a comma. Instead, find the last index of a comma and split it into two.

Reply:
Also you want to throw if there is no comma.